tracker/server.go

```diff
@@ -0,0 +1,237 @@
+package tracker
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	radio "github.com/R-a-dio/valkyrie"
+	"github.com/R-a-dio/valkyrie/config"
+	"github.com/R-a-dio/valkyrie/errors"
+	"github.com/R-a-dio/valkyrie/proxy/compat"
+	"github.com/R-a-dio/valkyrie/rpc"
+	"github.com/R-a-dio/valkyrie/util/graceful"
+	"github.com/R-a-dio/valkyrie/website"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/hlog"
+	"google.golang.org/grpc"
+)
+
+type Server struct {
+	cfg     config.Config
+	mu      sync.Mutex
+	rpcSrv  *grpc.Server
+	rpcLn   net.Listener
+	httpSrv *http.Server
+	httpLn  net.Listener
+}
+
+func NewGRPCServer(ctx context.Context, lts radio.ListenerTrackerService) *grpc.Server {
+	gs := rpc.NewGrpcServer(ctx)
+	rpc.RegisterListenerTrackerServer(gs, rpc.NewListenerTracker(lts))
+	return gs
+}
+
+func NewServer(ctx context.Context, cfg config.Config) (*Server, error) {
+	const op errors.Op = "tracker.NewServer"
+
+	// setup recorder
+	var recorder = NewRecorder(ctx, cfg)
+
+	// setup periodic task to update the manager of our listener count
+	go PeriodicallyUpdateListeners(ctx, cfg.Manager, recorder, UpdateListenersTickrate)
+	// setup periodic task to keep recorder state in sync with icecast
+	go PeriodicallySyncListeners(ctx, cfg, recorder, SyncListenersTickrate)
+
+	srv := new(Server)
+	r := website.NewRouter()
+
+	r.Use(
+		hlog.NewHandler(*zerolog.Ctx(ctx)),
+		hlog.RemoteAddrHandler("ip"),
+		hlog.UserAgentHandler("user_agent"),
+		hlog.RequestIDHandler("req_id", "Request-Id"),
+		hlog.URLHandler("url"),
+		hlog.MethodHandler("method"),
+		hlog.ProtoHandler("protocol"),
+		hlog.AccessHandler(zerologLoggerFunc),
+	)
+	r.Post("/listener_joined", ListenerAdd(ctx, recorder))
+	r.Post("/listener_left", ListenerRemove(ctx, recorder))
+
+	var err error
+
+	// setup the GRPC server that the rest will be poking
+	srv.rpcSrv = NewGRPCServer(ctx, recorder)
+	// and a listener for the GRPC server
+	srv.rpcLn, err = net.Listen("tcp", cfg.Conf().Tracker.RPCAddr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	srv.httpSrv = &http.Server{
+		Addr:         cfg.Conf().Tracker.ListenAddr,
+		Handler:      r,
+		ReadTimeout:  time.Second * 10,
+		WriteTimeout: time.Second * 5,
```

Comment:
Timeouts are probably not needed for this service, our client should be a well-behaving icecast so a timeout might instead actually cause us to lose requests.